### PR TITLE
wallet connect v2 adapter and provider

### PR DESCRIPTION
**Pull request by himanshuchawla009**

### CODE REVIEW

File: packages/adapters/wallet-connect-v2-adapter/src/walletConnectV2adapter.ts [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+  get provider(): SafeEventEmitterProvider | null {
+    return this.wcProvider?.provider || null;
+  }
+
+  set provider(_: SafeEventEmitterProvider | null) {
+    throw new Error("Not implemented");
+  }
+
+  async init(): Promise<void> {
```
Comment (arch1995) [inline packages/adapters/wallet-connect-v2-adapter/src/walletConnectV2adapter.ts:74]: Do we need to pass `autoConnect` param in init ?

Comment (himanshuchawla009) [reply packages/adapters/wallet-connect-v2-adapter/src/walletConnectV2adapter.ts]: resolved

File: packages/providers/ethereum-provider/src/providers/injectedProviders/WalletConnectV2Provider.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    engine.push(chainSwitchMiddleware);
+    engine.push(networkMiddleware);
+    const provider = providerFromEngine(engine);
+    this.updateProviderEngineProxy(provider);
+  }
+
+  private getChainSwitchMiddleware(): JRPCMiddleware<unknown, unknown> {
+    const chainSwitchHandlers: IChainSwitchHandlers = {
+      addChain: async (_: AddEthereumChainParameter): Promise<void> => {
+        // not required in wallet connect v2, since chains are added during connection.
```
Comment (chaitanyapotti) [inline packages/providers/ethereum-provider/src/providers/injectedProviders/WalletConnectV2Provider.ts:95]: wait.. web3auth doesn't accept multiple chains right in ctor.. we need to add this

File: packages/ui/src/components/Modal.tsx [hunk 3]
```diff
@@ -118,18 +118,28 @@ export default function Modal(props: ModalProps) {
...
-    handleExternalWalletClick(params);
-  };
+  const onExternalWalletClick = useCallback(
+    (params) => {
+      handleExternalWalletClick(params);
+    },
+    [handleExternalWalletClick]
+  );
+
+  const preHandleExternalWalletClick = useCallback(
```
Comment (chaitanyapotti) [inline packages/ui/src/components/Modal.tsx:128]: why this complicated syntax?

File: packages/providers/ethereum-provider/src/providers/injectedProviders/WalletConnectV2Provider.ts (hunk 2 above)
Comment (himanshuchawla009) [reply packages/providers/ethereum-provider/src/providers/injectedProviders/WalletConnectV2Provider.ts:95]: yes , it is required here forgot to add

**Review by chaitanyapotti — APPROVED** [review]
lgtm

**Pull request merged**